### PR TITLE
Add `match` CLI option to run only matching tests

### LINKED ISSUE #476: Only run tests with title matching expression
This is something that I would do with mocha. Here's the mocha help:

```
    -g, --grep <pattern>                    only run tests matching <pattern>
```

I'm hoping to get something like that working.

### PR DESCRIPTION
**Pull request by kasperlewau**
Add a `match` option for running a specific set of tests with titles matching the given pattern. Leverages [matcher](https://github.com/sindresorhus/matcher) internally, so `!` and `*` is supported. 

```
ava --match='foo*' --match='!*bar' /test/*.js

# shorthand
ava -m='foo*' -m='!*bar' /test/*.js
```

This would run all tests inside `/test` with a title starting with `foo` and not ending with `bar`. 

The match option can also be configured in the `ava` section of `package.json`. 

```json
{
  "ava": {
    "match": [
      "foo*"
    ]
  }
}
```

Fixes #476

### CODE REVIEW

**Comment by vdemedes** [comment]
I :heart: it! Simple code which does its job. Thank you!

There's one little thing though, could you please add some minimal docs to the readme along with a few examples?

**Comment by kasperlewau** [comment]
> There's one little thing though, could you please add some minimal docs to the readme along with a few examples?

Absolutely. I suppose somewhere in the `cli` section would make sense? Need to step away from open source for a couple of hours. Sometime tonight! :smile:

**Comment by vdemedes** [comment]
Yep, in `cli` section ;) Sure, no pressure, great job!

**Comment by kasperlewau** [comment]
Hm. Not sure why the build is failing, ran just fine™ on my machine. 

Node v5.7.0
Darwin v15.3.0

```
  options.match will skip tests with non-matching titles
  not ok TypeError: Cannot read property 'runOnlyExclusive' of undefined
```

**Comment by vdemedes** [comment]
All versions failed, so it must be something :)

**Comment by vdemedes** [comment]
Try pulling latest changes (#556 just got merged in).

**Comment by kasperlewau** [comment]
Nope. Still no sauce unfortunately. It does point in the direction of #566. 

Could it be that either `match` or `runOnlyExclusive` didn't add an entry in [`runner#chainableMethods`](https://github.com/sindresorhus/ava/pull/592/files#diff-b3b53682a18f203ac8d29b0e277cad26R9)? 

Stabbing in the dark here : )

**Comment by kasperlewau** [comment]
[this](https://github.com/sindresorhus/ava/pull/556/files#diff-b3b53682a18f203ac8d29b0e277cad26R99) appears to be the failing line. Somehow I'm not passing `options` to the `run` method.

File: lib/runner.js [hunk 1]
```diff
@@ -38,6 +39,7 @@ function Runner(options) {
 	this.tests = new TestCollection();
 	this._bail = options.bail;
 	this._serial = options.serial;
+	this._match = options.match;
```
Comment (novemberborn) [inline lib/runner.js]: These options must be explicitly forwarded in https://github.com/sindresorhus/ava/blob/4a1bd79fdd5dd13e011750b6cb2fefc94f74b5c6/index.js#L14

File: lib/runner.js [hunk 2]
```diff
@@ -55,6 +57,10 @@ optionChain(chainableMethods, function (opts, title, fn) {
 		opts.serial = true;
 	}
 
+	if (this._match) {
```
Comment (novemberborn) [inline lib/runner.js]: Once option forwarding is fixed you'll always have a value here. However you must make sure to only run the next line if `this._match` is not empty. By default it will be.

**Comment by novemberborn** [comment]
@kasperlewau nice work. However it doesn't actually work end-to-end. I've left some inline comments as to why. Could you add a test to `test/cli` to exercise the feature end-to-end?

**Comment by kasperlewau** [comment]
@vdemedes I've added docs for the `--match` flag (in the `cli` segment). We could probably shave off a sentence/example or two. 

I'm still on the fence regarding some wording. 
> match titles ending with 'foo' **[vs]** match titles that end with 'foo'

Let me know what you guys think.

**Comment by kasperlewau** [comment]
@novemberborn Wuh-oh! I'll see what I can do about that. Cheers!

**Comment by novemberborn** [comment]
> *Tests without a title will not run if* **any** *match pattern is supplied.*

Unfortunately the runner does not have access to the full test title. It's later [derived from the function name](https://github.com/sindresorhus/ava/blob/4a1bd79fdd5dd13e011750b6cb2fefc94f74b5c6/lib/test-collection.js#L45), for instance.

Maybe that title logic needs to be moved to the runner. @vdemedes?

File: lib/runner.js [hunk 3]
```diff
@@ -55,6 +57,10 @@ optionChain(chainableMethods, function (opts, title, fn) {
...
+		opts.skipped = matcher([title], this._match).length === 0;
```
Comment (novemberborn) [inline lib/runner.js]: This function is also called when adding hooks, which themselves can be titled. Would be weird to skip those. I think you should be able to guard against that by checking `opts.type === 'test'`.

Comment (kasperlewau) [inline lib/runner.js]: Right, because they too are a part of `chainableMethods`. willfix!

**Comment by kasperlewau** [comment]
@novemberborn 

* We're now passing through the `match` option from `api -> runner`.
* We're guarding against an empty `match` value. 
* We're guarding against non-`test` types in the 'apply-match-logic' if block. 

The one thing I haven't done is setting up an end-to-end test in `test/cli` to *ensure it all works™*. Do you have any suggestions/pointers I can take with me on that voyage? :smile:

**Comment by vdemedes** [comment]
@novemberborn It can't be moved to Runner, because we display full test titles only when there are multiple files and only CLI knows how many test files we have. And plus, it would be out of Runner's scope imho.

**Comment by kasperlewau** [comment]
@novemberborn @vdemedes 

I could reverse the behaviour by checking that title is not `null`, but to me as a user it'd feel weird to see tests without a title running when I've asked for a specific set of titles. 

The one time I feel a non-titled test should run, would be with either no match pattern supplied, or `--match=*`. But I don't see why one would ever want to supply such a pattern.

File: cli.js [hunk 4]
```diff
@@ -49,6 +49,7 @@ var cli = meow([
 	'  --tap, -t        Generate TAP output',
 	'  --verbose, -v    Enable verbose output',
 	'  --no-cache       Disable the transpiler cache',
+	'  --match, -m      Only run tests with matching title',
```
Comment (sindresorhus) [inline cli.js]: > '  --match, -m      Only run tests with matching title (Can be repeated)',

**Comment by novemberborn** [comment]
@vdemedes 

> It can't be moved to Runner, because we display full test titles only when there are multiple files and only CLI knows how many test files we have.

This isn't about test title prefixes. The runner calls `TestCollection#add()` which changes the title beyond what was determined in the runner, even though the runner has access to the same arguments.

File: readme.md [hunk 5]
```diff
@@ -134,6 +135,32 @@ Directories are recursive by default. Directories named `fixtures` and `helpers`
...
+$ ava --match=*foo
+
+# match titles starting with 'foo'
+$ ava --match=foo*
+
+# match titles containing 'foo'
+$ ava --match=*foo*
+
+# match titles not containing 'foo'
+$ ava --match=!*foo*
```
Comment (sindresorhus) [inline readme.md]: `!*foo*` needs to be quoted in ZSH at least.

**Comment by novemberborn** [comment]
> The one thing I haven't done is setting up an end-to-end test in test/cli to ensure it all works™. Do you have any suggestions/pointers I can take with me on that voyage?

See the "watcher works" test, though it does a few more things than are required here.

File: readme.md (hunk 5 above)
Comment (sindresorhus) [inline readme.md]: So does `foo*`, so probably better to quote them all?

Comment (kasperlewau) [inline readme.md]: Me being the inconsiderate Bash user, I opted for no quotes. I'll single quote 'em all :)

File: cli.js (hunk 4 above)
Comment (kasperlewau) [inline cli.js]: :+1:

**Comment by vdemedes** [comment]
@novemberborn Oh I see, I mixed that up. In that case, @kasperlewau's logic seems very reasonable:

> Tests without a title will not run if any match pattern is supplied.

> it'd feel weird to see tests without a title running when I've asked for a specific set of titles.

And still title manipulation feels out of runner's scope, so I'd leave it in test collection.

**Comment by novemberborn** [comment]
My point is that the runner calls another method which figures out an implicit title. Those are not matched. I guess it's OK to say that match only applies to explicit titles but it shouldn't be too hard to apply it to implicit ones too.

**Comment by vdemedes** [comment]
Ok, I see your point. What about merging this PR as it is and opening a new issue to extend this functionality? I see some refactoring work involved here.

**Comment by kasperlewau** [comment]
Ok, I believe that's the e2e test for `--match` in the bag. Fingers crossed Travis doesn't kick me in the groin : )

**Comment by sindresorhus** [comment]
I don't think it makes sense to match on implicit titles. It should IMHO only match on what people write out in their test file.

**Comment by novemberborn** [comment]
> What about merging this PR as it is and opening a new issue to extend this functionality? I see some refactoring work involved here.

Yes that's fair.

> I don't think it makes sense to match on implicit titles. It should IMHO only match on what people write out in their test file.

That's fair too. Not sure we want to encourage `test(function myTestTitle (t) {` style tests anyway.

Apologies for the distraction :smiley:

**Comment by kasperlewau** [comment]
I realised just now I took care of @sindresorhus's comment about repeatabilty - but I did so in the 'wrong' place :smile: 

I'll take care of that on my way to work!

**Comment by novemberborn** [comment]
Should this use `exclusive` rather than `skipped`? That is, set `exclusive` for matching tests. Perhaps in the Api report an error when no tests match. That makes `--match` similar to adding `.only` in the test files.

**Comment by kasperlewau** [comment]
> Should this use exclusive rather than skipped? That is, set exclusive for matching tests. 

Right off the bat that sounds better to me @novemberborn. I can't really comment further than that - I'm not familiar enough with the difference(s) between `skip` and `exclusive` when put into the context of say `--watch` or something of the sort. I bet you do though : ) 

If anything it would change the conditional statement to something slightly more pretty.

**Comment by novemberborn** [comment]
> I'm not familiar enough with the difference(s) between skip and exclusive when put into the context of say --watch or something of the sort. I bet you do though : )

It'll probably be OK, but we can make whatever improvements are needed once this lands.

> What I mean to say is that the semantics of match fit inclusion better than exclusion in my mind. Albeit that's a very fine line.

Here's another angle. If I have a `.only` test and I invoke AVA using a `--match`, what is the expected behavior? IMO `--match` should win (just like CLI flags override `package.json` config). That would be easier to achieve if the match implementation changes the `exclusive` field of the test metadata. Non-matching exclusive tests need to have that field set to `false`.

**Comment by kasperlewau** [comment]
> If I have a .only test and I invoke AVA using a `--match`, what is the expected behavior? IMO `--match` should win. 

Agreed. It's seldom that I pass CLI args to my test runner, and when I do pass them I expect them to be king of the hill.

I'm not so fussed about the easier way *(oh gosh, nitpicking...)* to achieve the end goal, but more so that the produced code is *simpler*. If switching to `exclusive` rather than `skip` to achieve CLI dominance would generate *simpler* code in the end, I'm all for it :+1:

**Comment by novemberborn** [comment]
I think it might make it easier to reason about the side-effects, e.g. in reporter output. Both `--match` and `.only` can cause a lot of tests not to run that would have run otherwise, unlike `.skip` which only causes skipped tests not to run.

**Comment by sindresorhus** [comment]
@novemberborn Does this still apply https://github.com/sindresorhus/ava/pull/592#discussion_r54730961 ?

**Comment by sindresorhus** [comment]
@novemberborn Is the `exclusive` change something you want to do in this PR or can it be a follow-up?

**Comment by novemberborn** [comment]
@sindresorhus,

> Does this still apply #592 (comment) ?

That's been resolved.

> Is the exclusive change something you want to do in this PR or can it be a follow-up?

I think we should do it here, shouldn't be too much work. As the PR stands now, if you have a `.only` test and use `--match` it will only be applied to `.only` tests. I think that's a bug.

**Comment by sindresorhus** [comment]
> if you have a .only test and use --match it will only be applied to .only tests. I think that's a bug.

Agreed.

**Comment by kasperlewau** [comment]
I've changed some stuff around to add the `exclusive` flag onto the given test if its title matches `--match`, leaving the `skipped` flag for `.skip`. 

As expected, non-matching tests no longer add to `testCount`, nor `skipCount`. That *could* open up for some confusion, as it's a fine line and a very personal opinion on whether a match pattern *excludes* or *includes*, but perhaps that is outside the scope what we want to achieve.

Should we introduce another counter to keep track of the tests that were not run due to the `--match` flag? `ignoreCount`? 

Sounds a lot like `skipCount` to me :unamused:, plus the fact that now we'd have to juggle even more words around; `skip, only, match, exclusive, ignore`. Gut feeling says that *could* be simplified, I'm just not quite sure how. 

--- 

Also, consider the following: 

`ava --match=*oo`
```js
test.skip('moo', t => {
  t.pass();
});

test('foo', t => {
  t.pass();
});
```

If `--match` takes precedence over `.only`, should it also override `.skip`?

**Comment by novemberborn** [comment]
> As expected, non-matching tests no longer add to testCount, nor skipCount. That could open up for some confusion, as it's a fine line and a very personal opinion on whether a match pattern excludes or includes, but perhaps that sort of balancing act is outside the scope of what we're trying to achieve here.
> 
> Should we introduce another counter to keep track of the tests that were not run due to the --match flag? ignoreCount?
> 
> Sounds a lot like skipCount to me , plus the fact that now we'd have to juggle even more words around; skip, only, match, exclusive, ignore. Gut feeling says that could be simplified, I'm just not quite sure how.

#471 covers this. Hopefully we can solve it in one fell swoop for all cases.

> If --match takes precedence over .only, should it also override .skip?

I'd argue no. You'd use `.skip` because a test is broken. You'd use `.only` or `--match` to select specific tests.

---

> I've changed some stuff around to add the exclusive flag onto the given test if its title matches --match, leaving the skipped flag for .skip.

Doesn't look like you pushed any commits yet though? No worries if you were waiting for feedback on your questions, but figured you might have forgotten ;-)

**Comment by kasperlewau** [comment]
> #471 covers this. Hopefully we can solve it in one fell swoop for all cases.

Gotcha. I mean something like that could become a mess quite quickly. Glad it's on the roadmap :) 

> I'd argue no. You'd use .skip because a test is broken. You'd use .only or --match to select specific tests.

I suppose you're right about that one. I was halfway through an explanation on why I feel it *should be so*, but halfway through I realised I ended up at the same conclusion - *no.* 

> Doesn't look like you pushed any commits yet though? No worries if you were waiting for feedback on your questions, but figured you might have forgotten ;-)

I'd say that was a 50/50. I was curious as to if it made sense from those point of views, and my net connection died on the train back home as I was about to push. Hadn't gotten around to it yet :smile:

**Comment by kasperlewau** [comment]
Rebased and good to go from my end. Not sure I nailed the `exclusive` flag, so I would love to get another checkup on this PR. 

Just waiting for those CI's to finish their business.

File: lib/runner.js [hunk 6]
```diff
@@ -69,6 +71,10 @@ optionChain(chainableMethods, function (opts, title, fn) {
 		opts.serial = true;
 	}
 
+	if (opts.type === 'test' && title !== null && (this._match && this._match.length > 0)) {
```
Comment (novemberborn) [inline lib/runner.js]: No need for the parentheses around the match check.

File: lib/runner.js [hunk 7]
```diff
@@ -99,7 +105,8 @@ Runner.prototype._addTestResult = function (result) {
 		error: result.reason,
 		type: test.metadata.type,
 		skip: test.metadata.skipped,
-		todo: test.metadata.todo
+		todo: test.metadata.todo,
+		exclusive: test.metadata.exclusive
```
Comment (novemberborn) [inline lib/runner.js]: Why is this necessary? Presumably `.only` already works without needing this prop.

File: readme.md [hunk 8]
```diff
@@ -134,6 +135,32 @@ Directories are recursive by default. Directories named `fixtures` and `helpers`
 
 When using `npm test`, you can pass positional arguments directly `npm test test2.js`, but flags needs to be passed like `npm test -- --verbose`.
 
+### The `--match` flag
```
Comment (novemberborn) [inline readme.md]: I think you should move this down below the `Only-tests` section. Maybe retitle to `Running test that match a pattern`. Should mention that this is more specific than `.only`.

File: readme.md [hunk 9]
```diff
@@ -134,6 +135,32 @@ Directories are recursive by default. Directories named `fixtures` and `helpers`
...
+
+The `--match` flag allows you to run a subset of the tests in your suite that have a matching title. This is achieved with simple wildcard patterns. For more information, check out [sindresorhus/matcher](https://github.com/sindresorhus/matcher).
```
Comment (novemberborn) [inline readme.md]: Drop the `sindresorhus/` part from the link title.

File: readme.md [hunk 10]
```diff
@@ -134,6 +135,32 @@ Directories are recursive by default. Directories named `fixtures` and `helpers`
...
+$ ava --match='!*foo*'
+
+# match titles starting with 'foo' and ending with 'bar'
+$ ava --match='foo*bar'
+
+# match titles starting with 'foo' or ending with 'bar'
+$ ava --match='foo*' --match='*bar'
+```
+
+*Tests without a title will not run if* **any** *match pattern is supplied.*
```
Comment (novemberborn) [inline readme.md]: Maybe "without an explicit title".

File: test/cli.js [hunk 11]
```diff
@@ -156,6 +156,13 @@ test('watcher works', function (t) {
 	});
 });
 
+test('--match works', function (t) {
+	execCli(['-m=foo', '-m=bar', '-m=!baz', '-m=t* a* f*', '-m=!t* a* n* f*', 'fixture/matcher-skip.js'], function (err) {
+		t.ifError(err);
```
Comment (novemberborn) [inline test/cli.js]: Nice!

**Comment by novemberborn** [comment]
LGTM! I'm in the midst of overhauling the readme (see #597) so don't worry too much about those changes.

File: lib/runner.js (hunk 7 above)
Comment (kasperlewau) [inline lib/runner.js]: It's not necessary. That's me mixing up two wholly different reasons for the full test suite failing. Removing in a jiffy!

File: lib/runner.js (hunk 6 above)
Comment (kasperlewau) [inline lib/runner.js]: Nope, not needed. This would be a case of me not killing my darlings. The amount of times I've been recommended to go and check out lisp is bordering on the ridiculous.

File: readme.md (hunk 8 above)
Comment (kasperlewau) [inline readme.md]: Sure thing!

File: readme.md (hunk 9 above)
Comment (kasperlewau) [inline readme.md]: :+1:

File: readme.md (hunk 10 above)
Comment (kasperlewau) [inline readme.md]: Yup, that's better.

File: test/cli.js (hunk 11 above)
Comment (kasperlewau) [inline test/cli.js]: : )

**Comment by kasperlewau** [comment]
@novemberborn I agree with the changes you proposed to the Readme. Would you like for me to handle those, or defer to #597?

File: readme.md (hunk 8 above)
Comment (kasperlewau) [inline readme.md]: Hm. How's about `Matched-tests`? Too succinct? 

Thought it'd look better to keep in line with the other 'test scope' modifiers; 
`Serial-tests`, `Only-tests`, `Skip-tests`, `Todo-tests`

**Comment by novemberborn** [comment]
> I agree with the changes you proposed to the Readme. Would you like for me to handle those, or defer to #597?

Seems fine as it is now. I'll give it another read within the context of the changes in #597 and adjust as necessary once this is in `master`.

Not sure why Travis failed in Node 5.

**Comment by kasperlewau** [comment]
> Not sure why Travis failed in Node 5.

Beats me :confused: Can we trigger a rerun and see what's up?

**Comment by novemberborn** [comment]
> Can we trigger a rerun and see what's up?

Just did. Didn't realize I had that power :muscle:

**Comment by kasperlewau** [comment]
> Just did. Didn't realize I had that power :muscle:

And with that, I will calmly await any further feedback/instructions. Flex them muscles, you re-runner you! 

Aside; I was halfway past throwing my MBP right out the window in my previous PR. Not having the power to trigger a re-run had me shouting in agony, being someone who loves to push with `-f` and pull with `--rebase`. 

:beer:

**Comment by sindresorhus** [comment]
> Not having the power to trigger a re-run had me shouting in agony

Just amend the latest commit without any new changes and force push and you trigger a rerun ;)

**Comment by kasperlewau** [comment]
> Just amend the latest commit without any new changes and force push and you trigger a rerun ;)

See I'm *quite sure* I did just that, but it simply wouldn't rebuild on Travis/AppVeyor, which struck me as odd as I am an avid wielder of the `--force`. Nevertheless, we be green now :smile:

**Comment by novemberborn** [comment]
@sindresorhus @vdemedes @jamestalmage LGTY?

**Comment by novemberborn** [comment]
Please make sure to rebase on `master` to [fix a linting issue](https://github.com/sindresorhus/ava/pull/601) and run the latest XO configuration against your PR. Thanks!

**Comment by kasperlewau** [comment]
> Please make sure to rebase on master to fix a linting issue and run the latest XO configuration against your PR. Thanks!

Done!

**Comment by sindresorhus** [comment]
LGTM2. Landed.

Superb work on this @kasperisager, and thanks for persevering through our extensive review. This turned out really good! :)

**Comment by kasperlewau** [comment]
My pleasure! And thanks :) 

Sidenote; Didn't get no notification. As it turns out, there actually *could be* too many Kasper's around @novemberborn @sindresorhus :smiley:!

facepalm
